tf2/src/buffer_core.cpp

```diff
@@ -115,52 +115,55 @@ std::string stripSlash(const std::string& in)
 }
 
 
-bool BufferCore::warnFrameId(const char* function_name_arg, const std::string& frame_id) const
+CompactFrameID BufferCore::validateFrameId(
+  const char* function_name_arg,
+  const std::string& frame_id,
+  bool should_throw,
+  std::string* error_msg) const
 {
-  if (frame_id.size() == 0)
+  std::stringstream ss;
+  CompactFrameID id(0);
+  
+  if (frame_id.empty())
   {
-    std::stringstream ss;
-    ss << "Invalid argument passed to "<< function_name_arg <<" in tf2 frame_ids cannot be empty";
-    CONSOLE_BRIDGE_logWarn("%s",ss.str().c_str());
-    return true;
+    ss << "Invalid argument passed to "<< function_name_arg << " - in tf2 frame_ids cannot be empty";
+    if (should_throw)
+    {
+      throw tf2::InvalidArgumentException(ss.str().c_str());
+    }
   }
-
-  if (startsWithSlash(frame_id))
+  else if (startsWithSlash(frame_id))
   {
-    std::stringstream ss;
-    ss << "Invalid argument \"" << frame_id << "\" passed to "<< function_name_arg <<" in tf2 frame_ids cannot start with a '/' like: ";
-    CONSOLE_BRIDGE_logWarn("%s",ss.str().c_str());
-    return true;
+    ss << "Invalid argument \"" << frame_id << "\" passed to "<< function_name_arg << " - in tf2 frame_ids cannot start with a '/'";
+    if (should_throw)
+    {
+      throw tf2::InvalidArgumentException(ss.str().c_str());
+    }
   }
-
-  return false;
-}
-
-CompactFrameID BufferCore::validateFrameId(const char* function_name_arg, const std::string& frame_id) const
-{
-  if (frame_id.empty())
+  else if ((id = lookupFrameNumber(frame_id)) == 0)
   {
-    std::stringstream ss;
-    ss << "Invalid argument passed to "<< function_name_arg <<" in tf2 frame_ids cannot be empty";
-    throw tf2::InvalidArgumentException(ss.str().c_str());
+    ss << "\"" << frame_id << "\" passed to "<< function_name_arg <<" does not exist.";
+    if (should_throw)
+    {
+      throw tf2::LookupException(ss.str().c_str());
+    }
   }
-
-  if (startsWithSlash(frame_id))
+  else
   {
-    std::stringstream ss;
-    ss << "Invalid argument \"" << frame_id << "\" passed to "<< function_name_arg <<" in tf2 frame_ids cannot start with a '/' like: ";
-    throw tf2::InvalidArgumentException(ss.str().c_str());
+    // No checks failed, this variable was populated in the prior lookup check
+    return id;
   }
 
-  CompactFrameID id = lookupFrameNumber(frame_id);
-  if (id == 0)
+  // A check failed and we fell through to here
+  if (error_msg)
   {
-    std::stringstream ss;
-    ss << "\"" << frame_id << "\" passed to "<< function_name_arg <<" does not exist. ";
-    throw tf2::LookupException(ss.str().c_str());
+    *error_msg = ss.str();
   }
-  
-  return id;
+  else
+  {
+    CONSOLE_BRIDGE_logWarn("%s", ss.str().c_str());
+  }
+  return 0;
```

Comment:
Very minor nit: worth defining an invalid frame constant?

Reply:
I think that's a good idea but it would require changing code outside the scope of this PR to be consistent, so I'm not inclined to do it here.